### PR TITLE
Loadtestservice - SDK initial version

### PR DESCRIPTION
**Pull request by Harshan01**
# Description
First time generated SDK for Azure Load Testing service (sdk/loadtestservice) using Swagger ([https://github.com/Azure/azure-rest-api-specs/blob/9401446f22177696d920cf110893a0de7452ee9e/specification/loadtestservice/data-plane/readme.md](https://github.com/Azure/azure-rest-api-specs/blob/9401446f22177696d920cf110893a0de7452ee9e/specification/loadtestservice/data-plane/readme.md))
* Auth using AADToken
* "multipart/form-data" file upload implemented manually using ByteStream.
* Has samples and tests for common champion scenario 
* Uses BinaryData for all methods.

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by Harshan01** [comment]
* What is the recommended approach for implementing multipart/form-data in SDK? Currently, external dependencies are not allowed, hence cannot use apache.httpcomponents.httpmime package for generating multipart/form-data request body.
* Autorest generates BinaryData for all method. Is it expected/recommended to replace them with autogenerated models after removing "@generated" decorator?

**Comment by Harshan01** [comment]
README, tests and samples will be updated after resolving the functionality issues caused by multipart/form-data and BinaryData requests. Commits will be squashed when PR is ready to be merged

**Comment by joshfree** [comment]
@samvaity is going to help shepherd this PR into the main branch

**Comment by Harshan01** [comment]
@samvaity Most of the changes regarding README, samples and test cases can be fixed once the functionality of the APIs are approved, since tests and samples would require changes if there are changes in client method names

File: sdk/loadtestservice/azure-developer-loadtesting/pom.xml
Comment (JonathanGiles) [reply sdk/loadtestservice/azure-developer-loadtesting/pom.xml:11]: Before anything gets released we need to ensure we have cross-language consensus. I can't recall if we have come to a consensus yet.

* I see on the C# side we appear to have `Azure.Analytics.LoadTestService`, 
* On the Java side we have both of `com.azure.developer.loadtesting` and `com.azure.analytics.loadtestservice`, and
* Python has both `azure.developer.loadtesting` and `azure.developer.loadtestservice`. 
 
This is a blocker for release - so if @Harshan01 can please clarify and ensure consistency before any beta releases happen, that would be appreciated.

Comment (Harshan01) [reply sdk/loadtestservice/azure-developer-loadtesting/pom.xml:11]: Sure, will update once we decide on naming

Comment (Harshan01) [reply sdk/loadtestservice/azure-developer-loadtesting/pom.xml:11]: @JonathanGiles @samvaity We will be going forward with `azure.developer.loadtesting` everywhere. Python currently does not have `azure.developer.loadtestservice`. All existing `azure.analytics.*` versions of packages will be removed from repos

File: sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md
Comment (weidongxu-microsoft) [inline sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md:13]: Not sure if you want to start this right now. This is only as reference, when you want to write convenient API and need to see what model would look like (but it might not what you want to use in API).

**Review by weidongxu-microsoft — COMMENTED** [review]
1. mutipart is not supported yet, old [issue](https://github.com/Azure/azure-sdk-for-java/issues/8702) @alzimmermsft 
2. DPG only generates protocol methods. You are welcome to add convenience APIs (usually, adding new API, not modifying the existing API annotated with `@Generated` unless they are implemented wrong). See ref below.

Ref https://github.com/Azure/azure-sdk-for-java/wiki/Protocol-Methods-Quickstart-with-AutoRest#improve-sdk-documentation

---

Added some comments to apiview. Quite some on "pageable" of "list" API.

And any of the API is long-running?

File: sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md
Comment (Harshan01) [reply sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md]: @weidongxu-microsoft @samvaity The PR is updated to be non-DPG. No API is officially long-running. There are a few APIs which internally functions like long-running but our service doesn't treat it as one

Comment (Harshan01) [reply sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md]: Yes, data-plane: true is not passed now, and pageable is added for required APIs

Comment (Harshan01) [reply sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md]: @weidongxu-microsoft @samvaity  in case we want to use DPG, would the users be expected to use BinaryData to interact with SDK? Or would models still need to be created? In case of DPG, is it fine to use partial update with models in place of BinaryData in protocol methods?

Comment (Harshan01) [reply sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md]: @weidongxu-microsoft Since multipart is not supported, is it fine to perform multipart upload manually as demonstrated in this commit?
https://github.com/Azure/azure-sdk-for-java/commit/7fcfb51efce5cd28602fd9a048821bdd426fd62d

Comment (weidongxu-microsoft) [reply sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md]: If release as DPG, user is expected to use BinaryData to interact with backend.

The point of DPG is that service can add model and convenience API (via partial update), after SDK released a few versions (even after a GA version).
Service is not required to have model or convenience API for preview or even for GA (though GA require much more test cases).

---

Yes, if payload is not large, one can use this approach to send multipart (I think I've tried that and it works, though likely one year ago).

**Comment by Harshan01** [comment]
@weidongxu-microsoft @samvaity Reverted back to DPG with samples and multipart upload. Test recordings are remaining

File: sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md
Comment (weidongxu-microsoft) [inline sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md:4]: we should only use spec from Azure repo. Make sure you update to Azure repo before merge.

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/TestAsyncClient.java
Comment (weidongxu-microsoft) [inline sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/TestAsyncClient.java]: If the content is from `BinaryData`, why user need to provide a `filename`? The data might just from memory?

---

Method signature.

I think the handwritten overload should be without the last (`RequestOptions` param).

And the sync one should have `Context` param (in place of `RequestOptions`).

Comment (Harshan01) [reply sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/TestAsyncClient.java]: `BinaryData.fromFile` discards the filename. You can refer the "Uploading .jmx file to Load Test" section in the README sample. Is there an alternate approach to preserve filename in `BinaryData` while keeping user experience simple?

Comment (Harshan01) [reply sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/TestAsyncClient.java]: We would require `RequestOptions` parameter in Async and Sync because `filetype` is an optional query param which needs to be passed via `RequestOptions`

Comment (weidongxu-microsoft) [reply sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/TestAsyncClient.java]: If you only support upload from file, you should use a `File` parameter, instead of `BinaryData` parameter.

---

For optional parameter, the standard "WithResponse" API would have all required/optional parameters in method signature (user can choose to use `null` if they don't want to fill).

And you can have another e.g. "uploadTestFile" method with only required parameters.

---

Ref, guideline https://azure.github.io/azure-sdk/java_introduction.html

**Comment by Harshan01** [comment]
@samvaity @weidongxu-microsoft Test recordings for champion scenario are added. Requesting review

File: sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md
Comment (weidongxu-microsoft) [inline sdk/loadtestservice/azure-developer-loadtesting/swagger/README.md:4]: the file does not have `x-ms-pageable`?

**Comment by Harshan01** [comment]
@samvaity  @weidongxu-microsoft Changes for administration subclients have been done. Requesting review. However, the pipelines fail even though test recordings are present. Is there anything that needs to be done to make the pipelines pass (except the broken links which would be fixed when PR is merged)

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/TestRunClientBuilder.java
Comment (weidongxu-microsoft) [inline sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/TestRunClientBuilder.java:44]: We seems no longer need this 2 Builders?

**Review by weidongxu-microsoft — COMMENTED** [review]
Still some comments pending.

We might first get the client/subclient reviewed, before getting to others including tests (on a quick look I don't see why it fails, and it was good if I remember correctly).

@samvaity Do we need one approval from sdk archboard on the client/subclient API?

**Comment by Harshan01** [comment]
@weidongxu-microsoft @samvaity APIView link for the current version of SDK: https://apiview.dev/Assemblies/Review/95fbc680963e4a40ad0edc4c956357c4

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/implementation/util/MultipartHelper.java
Comment (samvaity) [inline sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/implementation/util/MultipartHelper.java:22]: It would be good to add documentation to the method.
cc: @vcolin7

File: sdk/loadtestservice/azure-developer-loadtesting/src/test/resources/sample-JMX-file.jmx
Comment (samvaity) [inline sdk/loadtestservice/azure-developer-loadtesting/src/test/resources/sample-JMX-file.jmx:1]: delete file?

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationClient.java
Comment (samvaity) [inline sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationClient.java:93]: This should have `context` param.
https://azure.github.io/azure-sdk/java_introduction.html#java-service-client-context

Every API should have a simple method without options/context and another `withResponse` overload for options with context param.
Here and below.

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationAsyncClient.java
Comment (samvaity) [inline sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationAsyncClient.java:194]: Missing deleteAppComponentWithResponse?

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationAsyncClient.java
Comment (samvaity) [inline sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationAsyncClient.java:321]: Missing getAppComponent?

**Comment by Harshan01** [comment]
Latest APIView: https://apiview.dev/Assemblies/Review/95fbc680963e4a40ad0edc4c956357c4?revisionId=172bab5d058e46bcacb3f712c6802c72&doc=False

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationAsyncClient.java
Comment (weidongxu-microsoft) [inline sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationAsyncClient.java:741]: If you didn't expect user to call this method, remove `@Generated` and change the method to package private. Same to the method in sync client.

**Comment by Harshan01** [comment]
APIView: https://apiview.dev/Assemblies/Review/95fbc680963e4a40ad0edc4c956357c4?revisionId=1fed57fa4c0245518738cb667b84cd6b&doc=False

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestingAsyncClient.java
Comment (samvaity) [inline sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestingAsyncClient.java:38]:
```suggestion
    public LoadTestAdministrationAsyncClient getLoadTestAdministrationAsyncClient?() {
```

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestingAsyncClient.java
Comment (samvaity) [inline sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestingAsyncClient.java:47]:
```suggestion
    public TestRunAsyncClient getLoadTestRunAsyncClient() {
```

Comment (samvaity) [reply sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestingAsyncClient.java:47]: same for sync client

File: sdk/loadtestservice/azure-developer-loadtesting/CHANGELOG.md
Comment (samvaity) [inline sdk/loadtestservice/azure-developer-loadtesting/CHANGELOG.md:7]: Missing info on entries, can be done in a follow up PR.

**Review by samvaity — APPROVED** [review]
Changes lgtm, subclient design can be followed up post this preview release.

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationClient.java
Comment (samvaity) [inline sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationClient.java:971]: Was this change handwritten?

**Comment by Harshan01** [comment]
APIView: https://apiview.dev/Assemblies/Review/95fbc680963e4a40ad0edc4c956357c4?revisionId=0bdf51079291442eb9bb393757f88747&doc=False

File: sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationClient.java
Comment (weidongxu-microsoft) [reply sdk/loadtestservice/azure-developer-loadtesting/src/main/java/com/azure/developer/loadtesting/LoadTestAdministrationClient.java:971]: This is generated. I think Harshan is addressing my comment here that name should follow `list<noun>`.
https://azure.github.io/azure-sdk/java_introduction.html#java-service-client-verbs

Fine for me. Though usually you should sync with other SDKs and check if you should change the swagger directly.

File: sdk/loadtestservice/azure-developer-loadtesting/README.md
Comment (weidongxu-microsoft) [reply sdk/loadtestservice/azure-developer-loadtesting/README.md:48]: This sample is not correct. `AzureProfile` is from azure-core-management. User would not use this lib in data-plane.

Please remove it, as well as the line `    .authorityHost(profile.getEnvironment().getActiveDirectoryEndpoint())` (default host should work for Azure Global).

As Sameeksha mentioned, you'd better to put it to a file and use tag to insert snippet (this would let you avoid this issue mentioned above).

File: sdk/loadtestservice/azure-developer-loadtesting/README.md
Comment (weidongxu-microsoft) [inline sdk/loadtestservice/azure-developer-loadtesting/README.md:1]: make sure you've run `npx cspell` on the README.md, and fix spelling error if found.

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
Pending to be resolved before GA

1. method name "createAndUpdateTestRun"
2. method signature "uploadTestFile" (should we use `File file` instead of `String fileName, BinaryDate file`)
3. client/subclient structure

File: sdk/loadtestservice/azure-developer-loadtesting/README.md
Comment (ninallam) [inline sdk/loadtestservice/azure-developer-loadtesting/README.md:1]: Azure Load Testing client library for Java

File: sdk/loadtestservice/azure-developer-loadtesting/README.md
Comment (ninallam) [inline sdk/loadtestservice/azure-developer-loadtesting/README.md:137]: Can we add sample data for secrets and pass fail criteria as well?

File: sdk/loadtestservice/azure-developer-loadtesting/README.md
Comment (ninallam) [inline sdk/loadtestservice/azure-developer-loadtesting/README.md:116]: Please add some comments through out the sample to make it readable.

File: sdk/loadtestservice/azure-developer-loadtesting/README.md
Comment (ninallam) [inline sdk/loadtestservice/azure-developer-loadtesting/README.md:178]: Can we add a  sample code to fetch the client metrics for this test run to complete the end-to-end scenario?

**Pull request merged**
